content/writing_tests/migration-sakuli-1.md

```diff
@@ -0,0 +1,168 @@
+
+Back then, the most common way to handle asynchronous code was via callbacks. They were passed on to the asynchronous operation where they were executed as soon as the initial operation was finished. This approach was useful and acceptable for the usual JavaScript use-case during those days: Handling events in the DOM. But when Node was released and many browser applications became more and more complex, the developers ended up with a so called <a href="http://callbackhell.com/" target="_blank">callback hell</a>. To exit this callback hell, ES6 introduced the <a href="https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Promise" target="_blank">Promise class</a>. The advantage of a Promise is its ability to chain with other Promises, making it way more elegant to write subsequent, dependent asynchronous operations. A good read on this topic can be found on chapter VIII of <a href="https://exploringjs.com/impatient-js/toc.html" target="_blank">JavaScript for impatient programmers</a>
+
+#### The async wrapper
+
+You might have noticed that if a testcase in Sakuli v2 is wrapped within `(async () => /*...*/)().then(done)` this construct it is called *Immediately Invoked Function Expression* (IIFE). Basically, it is the definition of a function that is immediately invoked. This pattern is widely used in JavaScript to preserve scopes and namespaces, since every symbol, which is defined at the top level of a script, becomes part of the global namespace and may potentially collide with other scripts. In the era of <a href="https://medium.com/fusebox/beginner-web-developers-use-a-bundler-31ab0c91d2f5" target="_blank">web bundler</a> and <a href="https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/" target="_blank">ES-modules</a> this use-case became more and more irrelevant. IIFE on the other hand still remains important after the introduction of async/await in JavaScript. One caveat of this syntax is, that the `await` keyword is only available within an asynchronous function - because the engine will wrap the whole body of this function within a Promise. That means, that in order to be able to make use of the `await` keyword, your code has to be executed within an asynchronous function. On the top level IIFE, we declare the actual function as `async` and can make use of `await`. Since the result of this function is automatically turned into a Promise, we can invoke `then` and pass the `done` function to it. `done` is a Sakuli specific function, which indicates to Sakuli that the testcase script is finished. There is currently a discussion on the <a href="https://github.com/tc39/proposal-top-level-await" target="_blank">JavaScript proposal in stage 3</a>, which would introduce a *top level await* in JavaScript.
```

Comment:
```suggestion
You might have noticed that if a testcase in Sakuli v2 is wrapped within `(async () => /*...*/)()` this construct it is called *Immediately Invoked Function Expression* (IIFE). Basically, it is the definition of a function that is immediately invoked. This pattern is widely used in JavaScript to preserve scopes and namespaces, since every symbol, which is defined at the top level of a script, becomes part of the global namespace and may potentially collide with other scripts. In the era of <a href="https://medium.com/fusebox/beginner-web-developers-use-a-bundler-31ab0c91d2f5" target="_blank">web bundler</a> and <a href="https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/" target="_blank">ES-modules</a> this use-case became more and more irrelevant. IIFE on the other hand still remains important after the introduction of async/await in JavaScript. One caveat of this syntax is, that the `await` keyword is only available within an asynchronous function - because the engine will wrap the whole body of this function within a Promise. That means, that in order to be able to make use of the `await` keyword, your code has to be executed within an asynchronous function. On the top level IIFE, we declare the actual function as `async` and can make use of `await`.
```